File: solutions/0036. Valid Sudoku/0036.py
# Time complexity: O(1)
# Space complexity: O(1)
class Solution:
  def isValidSudoku(self, board: List[List[int]]) -> bool:
    seen = set()

    for i in range(len(board)):
      for j in range(len(board[0])):
        c = board[i][j]
        if c == '.':
          continue
        if c + '@row' + str(i) in seen or \
           c + '@col' + str(j) in seen or \
           c + '@box' + str(i // 3) + str(j // 3) in seen:
          return False
        seen.add(c + '@row' + str(i))
        seen.add(c + '@col' + str(j))
        seen.add(c + '@box' + str(i // 3) + str(j // 3))

    return True    

File: solutions/0036. Valid Sudoku/0036.cpp
// Time complexity: O(1)
// Space complexity: O(1)
class Solution {
 public:
  bool isValidSudoku(const vector<vector<char>>& board) {
    unordered_set<string> seen;

    for (int i = 0; i < board.size(); ++i)
      for (int j = 0; j < board[0].size(); ++j) {
        if (board[i][j] == '.')
          continue;
        const string c(1, board[i][j]);
        if (!seen.insert(c + "@row" + to_string(i)).second ||
            !seen.insert(c + "@col" + to_string(j)).second ||
            !seen.insert(c + "@box" + to_string(i / 3) + to_string(j / 3))
            .second)
          return false;
      }

    return true;
  }
};